### ISSUE TITLE
http panic while doing a network delete 

### ISSUE THREAD

**Issue by vvb** [contiv/netplugin#208]
netmaster, logs:
https://gist.github.com/68c5730057feac0a1a45

complete logs:
https://gist.github.com/a0f6fab194b77782cb7e

backtrace:
time="2015-12-14T14:57:40Z" level=info msg="Sending DELETE endpoint: &{EndpointID:192.168.24.1 EndpointType:internal EndpointGroup:0 IpAddr:192.168.24.1 VrfId:0 MacAddrStr:02:02:c0:a8:18:01 Vlan:1000 Vni:1000 OriginatorIp:192.168.2.10 PortNo:2 Timestamp:2015-12-14 14:45:22.446555596 +0000 UTC} to node 10.0.2.15:9002"
time="2015-12-14T14:57:40Z" level=info msg="Deleting docker network: private"
time="2015-12-14T14:57:40Z" level=info msg="freeing vlan 1000 vxlan 1000"
2015/12/14 14:57:40 http: panic serving [::1]:41979: runtime error: index out of range
goroutine 58 [running]:
net/http.(_conn).serve.func1(0xc820096210, 0x7f0838f5e3f8, 0xc820108020)
/usr/local/go/src/net/http/server.go:1287 +0xb5
github.com/jainvipin/bitset.(_BitSet).Set(0xc82046d500, 0xffffffffffffffff, 0xc82049f7e0)
/opt/gopath/src/github.com/contiv/netplugin/Godeps/_workspace/src/github.com/jainvipin/bitset/bitset.go:123 +0x163
github.com/contiv/netplugin/resources.(_AutoVXLANCfgResource).Deallocate(0xc8203dda10, 0x991b80, 0xc82049f7e0, 0x0, 0x0)
/opt/gopath/src/github.com/contiv/netplugin/resources/vxlanresource.go:173 +0x1e2
github.com/contiv/netplugin/resources.(_StateResourceManager).DeallocateResourceVal(0xc8201029e0, 0xc82046a830, 0x7, 0xa56660, 0xa, 0x991b80, 0xc82049f7e0, 0x0, 0x0)
/opt/gopath/src/github.com/contiv/netplugin/resources/stateresourcemanager.go:187 +0x311
github.com/contiv/netplugin/gstate.(_Cfg).FreeVXLAN(0xc8200c18c0, 0x7f0838f9e4f8, 0xc8201029e0, 0x3e8, 0x3e8, 0x0, 0x0)
/opt/gopath/src/github.com/contiv/netplugin/gstate/gstate.go:305 +0x1b1
github.com/contiv/netplugin/netmaster/master.freeNetworkResources(0x7f0838f5e318, 0xc820108090, 0xc8200d8750, 0xc8200c18c0, 0x0, 0x0)
/opt/gopath/src/github.com/contiv/netplugin/netmaster/master/network.go:463 +0x5ec
github.com/contiv/netplugin/netmaster/master.DeleteNetworkID(0x7f0838f5e318, 0xc820108090, 0xc82049e190, 0xf, 0x0, 0x0)
/opt/gopath/src/github.com/contiv/netplugin/netmaster/master/network.go:515 +0x791
github.com/contiv/netplugin/netmaster/objApi.(_APIController).NetworkDelete(0xc820108098, 0xc820159600, 0x0, 0x0)
/opt/gopath/src/github.com/contiv/netplugin/netmaster/objApi/apiController.go:453 +0x37f
github.com/contiv/objmodel/contivModel.DeleteNetwork(0xc82045c1f5, 0xf, 0x0, 0x0)
/opt/gopath/src/github.com/contiv/netplugin/Godeps/_workspace/src/github.com/contiv/objmodel/contivModel/contivModel.go:1446 +0x3cc
github.com/contiv/objmodel/contivModel.httpDeleteNetwork(0x7f0838f9ee38, 0xc8200962c0, 0xc8200de1c0, 0xc82045c2d0, 0x0, 0x0, 0x0, 0x0)
/opt/gopath/src/github.com/contiv/netplugin/Godeps/_workspace/src/github.com/contiv/objmodel/contivModel/contivModel.go:1365 +0x153
github.com/contiv/objmodel/contivModel.makeHttpHandler.func1(0x7f0838f9ee38, 0xc8200962c0, 0xc8200de1c0)
/opt/gopath/src/github.com/contiv/netplugin/Godeps/_workspace/src/github.com/contiv/objmodel/contivModel/contivModel.go:445 +0x73
net/http.HandlerFunc.ServeHTTP(0xc8201a0ce0, 0x7f0838f9ee38, 0xc8200962c0, 0xc8200de1c0)
/usr/local/go/src/net/http/server.go:1422 +0x3a
github.com/gorilla/mux.(_Router).ServeHTTP(0xc820116320, 0x7f0838f9ee38, 0xc8200962c0, 0xc8200de1c0)
/opt/gopath/src/github.com/contiv/netplugin/Godeps/_workspace/src/github.com/gorilla/mux/mux.go:98 +0x29e
net/http.serverHandler.ServeHTTP(0xc820139bc0, 0x7f0838f9ee38, 0xc8200962c0, 0xc8200de1c0)
/usr/local/go/src/net/http/server.go:1862 +0x19e
net/http.(_conn).serve(0xc820096210)
/usr/local/go/src/net/http/server.go:1361 +0xbee
created by net/http.(*Server).Serve
/usr/local/go/src/net/http/server.go:1910 +0x3f6

**Comment by jainvipin** [comment]
@vijayvikrant - is this fixed? I assume the steps to reproduce are:
- create network
- add containers to the network
- delete network
- stop container

Can you please add a test case for this?

**Comment by jainvipin** [comment]
I think there are chain of errors, that needs some attention. Ping @shaleman, @vijayvikrant @DivyaVavili 

With above mentioned steps, netplugin asserts (assuming the resource allocation changes would fix all that). I kept continuing with things to see if I could recover:
- In the step4 above, the container exit/stop gets stuck
- So, I Ctrl+C to come out, however this leaves with vport1 still there
- Now, if I restart netplugin/netmaster (using startPlugin.py) - they come up. However network creation fails because skydns tries to attach to the network and vport1 is already present
-  To continue going with this, if I create the network again, it works. It was a bit surprising initially, but the reason is that during previous failure the cleanup didn't clean and so vport1 is technically considered consumed. In fact we can `etcdctl get /contiv.io/state/nets/<net-name>` to see that eps attached to the network are not removed
- To continue further, if I now start a container, it also fails for the same reason (the error is bizzare, need another bug for this, which says "file already exists")
- Then again, after restarting that works too for the same reason

Couple of things we need to take care of:
- Failure reason should be clear, so user can go manually delete the vport1 in the host
- I am not sure if this is the right thing to do, but withdrawing what we did earlier upon failure would be something we should do

We certainly need to  try these things after resource allocation is moved to be done globally. However, I do think that opening up issues for some of the issues would be sensible in order for us to not get into unpredictable behavior and/or resource leaking.
